types/index.d.ts

```diff
@@ -1482,6 +1482,26 @@ declare module 'mongoose' {
 
   type ProjectionFields<DocType> = { [Key in keyof Omit<LeanDocument<DocType>, '__v'>]?: any } & Record<string, any>;
 
+  type _QueryManyForModel<T, ResultDoc, TQueryHelpers = {}> =
+    QueryWithHelpers<Array<ResultDoc>, ResultDoc, TQueryHelpers, T>;
+  type _QueryOneForModel<T, ResultDoc, TQueryHelpers = {}> =
+    QueryWithHelpers<ResultDoc, ResultDoc, TQueryHelpers, T>;
+
+  /**
+   * Helper type for getting a find Query type for a given Model (return of Model.find)
+   */
+  type QueryForModel<T extends Model<any>> =
+    T extends Model<infer MT, infer TQueryHelpers, infer TMethodsAndOverrides, infer TVirtuals> ?
```

Comment:
I'm very wary of using `infer` because it has caused some serious performance issues for us in the past. Can you please provide an example of why this type is helpful?

Reply:
This is just a helper utility type, so you can always not use it if it causes performance issues. Example of use would just be any time you need to return a query or create one but want to assign it to a variable declared earlier.

How I was using it:

```typescript

function applyQueryRestrictions(userQ: mongoose.QueryForModel<typeof UserModel>, requiredPerm?: string) {
    if (requiredPerm) {
        userQ.where('permissions', requiredPerm);
    }
    userQ.select('passInvalid username call first last email modified permissions')
        .select('verifiedCall.callsign verifiedCall.displayName verifiedCall.sigdate')
        .select('providers.type');
    userQ.populate({
        path: 'veDocs',
        select: 'isTeamLead active deleted number display_name vec call',
        populate: {
            path: 'ulsRecord',
            select: 'callsign frn license_class expired_date',
        }
    });
    return userQ;
}

```

I certainly could have constructed the query differently, but I wanted to keep type checking as much as possible and it's much easier to do  `QueryForModel<MyModel>` than it is to do `QueryWithHelpers<ResultDoc[], ResultDoc, {}, DocType>` -- not to mention that you have to figure out what all of those are. My original plan was to just do `type MyQueryType = ReturnType<typeof MyModel.find>` but since the return type of the different `.find` overrides aren't all identical that doesn't return a useful type.

Since this is just a utility method I can pull it and put it in my own code -- just seemed like something that others would also find useful and since it's purely optional it didn't seem like something which would cost anything to put in. Your call, just let me know.